File: src/commonMain/kotlin/com/scoreo/application/SyncUseCase.kt
package com.scoreo.application

import com.scoreo.domain.port.CloudSyncRepository
import kotlinx.datetime.toLocalDateTime
import com.scoreo.domain.port.GameTypeRepository
import com.scoreo.domain.port.MatchRepository
import com.scoreo.domain.port.PlayerRepository
import com.scoreo.domain.port.SyncData
import com.scoreo.domain.port.SyncException
import com.scoreo.domain.port.SyncStatus

data class SyncSnapshot(
    val playerCount: Int,
    val matchCount: Int,
    val gameTypeCount: Int,
    val lastModified: Long,
    val dateLabel: String?,
)

data class SyncConflict(
    val localSnapshot: SyncSnapshot,
    val remoteSnapshot: SyncSnapshot,
)

data class SyncResult(
    val pushed: Int = 0,
    val pulled: Int = 0,
    val timestamp: Long,
)

sealed class SyncOutcome {
    data class Synced(val result: SyncResult) : SyncOutcome()
    data class Conflict(val conflict: SyncConflict) : SyncOutcome()
}

class SyncUseCase(
    private val cloudRepo: CloudSyncRepository,
    private val playerRepo: PlayerRepository,
    private val gameTypeRepo: GameTypeRepository,
    private val matchRepo: MatchRepository,
) {
    fun autoSync(): SyncOutcome {
        val localData = buildLocalSyncData()
        val remoteData = cloudRepo.pull()

        val localEmpty = localData.players.isEmpty() && localData.gameTypes.isEmpty() && localData.matches.isEmpty()
        val remoteEmpty = remoteData.players.isEmpty() && remoteData.gameTypes.isEmpty() && remoteData.matches.isEmpty()

        when {
            localEmpty && remoteEmpty -> return SyncOutcome.Synced(SyncResult(timestamp = now()))
            remoteEmpty -> {
                cloudRepo.push(localData)
                return SyncOutcome.Synced(
                    SyncResult(
                        pushed = localData.players.size + localData.gameTypes.size + localData.matches.size,
                        timestamp = now(),
                    )
                )
            }
            localEmpty -> {
                writeRemoteToLocal(remoteData)
                return SyncOutcome.Synced(
                    SyncResult(
                        pulled = remoteData.players.size + remoteData.gameTypes.size + remoteData.matches.size,
                        timestamp = now(),
                    )
                )
            }
            else -> {
                if (isSameState(localData, remoteData)) {
                    return SyncOutcome.Synced(SyncResult(timestamp = now()))
                }
                return SyncOutcome.Conflict(
                    SyncConflict(
                        localSnapshot = toSnapshot(localData),
                        remoteSnapshot = toSnapshot(remoteData),
                    )
                )
            }
        }
    }

    fun resolveConflict(keepLocal: Boolean): SyncResult {
        if (keepLocal) {
            val localData = buildLocalSyncData()
            cloudRepo.push(localData)
            return SyncResult(
                pushed = localData.players.size + localData.gameTypes.size + localData.matches.size,
                timestamp = now(),
            )
        } else {
            val remoteData = cloudRepo.pull()
            writeRemoteToLocal(remoteData)
            return SyncResult(
                pulled = remoteData.players.size + remoteData.gameTypes.size + remoteData.matches.size,
                timestamp = now(),
            )
        }
    }

    fun login() {
        cloudRepo.login()
    }

    fun logout() {
        cloudRepo.logout()
    }

    fun status(): SyncStatus = cloudRepo.getStatus()

    // ── Private ──

    private fun buildLocalSyncData(): SyncData {
        val players = playerRepo.getAll(includeInactive = true)
        val gameTypes = gameTypeRepo.getAll()
        val matches = matchRepo.getAll()
        return SyncData(
            players = players,
            gameTypes = gameTypes,
            matches = matches,
            lastModified = now(),
        )
    }

    private fun writeRemoteToLocal(data: SyncData) {
        data.players.forEach { playerRepo.save(it) }
        data.gameTypes.forEach { gameTypeRepo.save(it) }
        data.matches.forEach { matchRepo.save(it) }
    }

    private fun isSameState(a: SyncData, b: SyncData): Boolean {
        return a.players.size == b.players.size
            && a.gameTypes.size == b.gameTypes.size
            && a.matches.size == b.matches.size
            && a.lastModified == b.lastModified
    }

    private fun toSnapshot(data: SyncData): SyncSnapshot = SyncSnapshot(
        playerCount = data.players.size,
        matchCount = data.matches.size,
        gameTypeCount = data.gameTypes.size,
        lastModified = data.lastModified,
        dateLabel = formatDate(data.lastModified),
    )

    private fun formatDate(epochMs: Long): String? {
        if (epochMs <= 0L) return null
        return try {
            val instant = kotlinx.datetime.Instant.fromEpochMilliseconds(epochMs)
            val local = instant.toLocalDateTime(kotlinx.datetime.TimeZone.currentSystemDefault())
            "${local.dayOfMonth} ${local.month.name.lowercase().take(3)} ${local.year}"
        } catch (_: Exception) { null }
    }

    companion object {
        private fun now(): Long =
            kotlinx.datetime.Clock.System.now().toEpochMilliseconds()
    }
}

File: src/application/syncUseCase.ts
import type {
  CloudSyncRepository,
  SyncData,
  SyncStatus,
} from '../domain/port/cloudSyncRepository'
import type { GameTypeRepository } from '../domain/port/gameTypeRepository'
import type { MatchRepository } from '../domain/port/matchRepository'
import type { PlayerRepository } from '../domain/port/playerRepository'

export interface SyncSnapshot {
  playerCount: number
  matchCount: number
  gameTypeCount: number
  lastModified: number
  dateLabel: string | null
}

export interface SyncConflict {
  localSnapshot: SyncSnapshot
  remoteSnapshot: SyncSnapshot
}

export interface SyncResult {
  pushed: number
  pulled: number
  timestamp: number
}

export type SyncOutcome = { kind: 'Synced'; result: SyncResult } | { kind: 'Conflict'; conflict: SyncConflict }

function now(): number {
  return Date.now()
}

function formatDate(epochMs: number): string | null {
  if (epochMs <= 0) return null
  try {
    const date = new Date(epochMs)
    const day = date.getDate()
    const month = date.toLocaleString('en-US', { month: 'short' }).toLowerCase()
    const year = date.getFullYear()
    return `${day} ${month} ${year}`
  } catch {
    return null
  }
}

function sortedById<T extends { id: string }>(items: T[]): T[] {
  return [...items].sort((a, b) => a.id.localeCompare(b.id))
}

/**
 * Structural equality independent of object key order, matching Kotlin data class `==`.
 * Plain `JSON.stringify` comparison is key-order-sensitive, which would report a false
 * Conflict for objects that are semantically identical but serialized in a different
 * field order (e.g. after a round-trip through a real network adapter's JSON parsing).
 */
function stableStringify(value: unknown): string {
  if (Array.isArray(value)) {
    return `[${value.map(stableStringify).join(',')}]`
  }
  if (value !== null && typeof value === 'object') {
    const record = value as Record<string, unknown>
    const keys = Object.keys(record).sort()
    return `{${keys.map((key) => `${JSON.stringify(key)}:${stableStringify(record[key])}`).join(',')}}`
  }
  return JSON.stringify(value)
}

function isSameState(a: SyncData, b: SyncData): boolean {
  return (
    stableStringify(sortedById(a.players)) === stableStringify(sortedById(b.players)) &&
    stableStringify(sortedById(a.gameTypes)) === stableStringify(sortedById(b.gameTypes)) &&
    stableStringify(sortedById(a.matches)) === stableStringify(sortedById(b.matches))
  )
}

function toSnapshot(data: SyncData): SyncSnapshot {
  return {
    playerCount: data.players.length,
    matchCount: data.matches.length,
    gameTypeCount: data.gameTypes.length,
    lastModified: data.lastModified,
    dateLabel: formatDate(data.lastModified),
  }
}

export class SyncUseCase {
  constructor(
    private readonly cloudRepo: CloudSyncRepository,
    private readonly playerRepo: PlayerRepository,
    private readonly gameTypeRepo: GameTypeRepository,
    private readonly matchRepo: MatchRepository,
  ) {}

  async autoSync(): Promise<SyncOutcome> {
    const localData = this.buildLocalSyncData()
    const remoteData = await this.cloudRepo.pull()

    const localEmpty = localData.players.length === 0 && localData.gameTypes.length === 0 && localData.matches.length === 0
    const remoteEmpty = remoteData.players.length === 0 && remoteData.gameTypes.length === 0 && remoteData.matches.length === 0

    if (localEmpty && remoteEmpty) {
      return { kind: 'Synced', result: { pushed: 0, pulled: 0, timestamp: now() } }
    }
    if (remoteEmpty) {
      await this.cloudRepo.push(localData)
      return {
        kind: 'Synced',
        result: {
          pushed: localData.players.length + localData.gameTypes.length + localData.matches.length,
          pulled: 0,
          timestamp: now(),
        },
      }
    }
    if (localEmpty) {
      this.writeRemoteToLocal(remoteData)
      return {
        kind: 'Synced',
        result: {
          pushed: 0,
          pulled: remoteData.players.length + remoteData.gameTypes.length + remoteData.matches.length,
          timestamp: now(),
        },
      }
    }
    if (isSameState(localData, remoteData)) {
      return { kind: 'Synced', result: { pushed: 0, pulled: 0, timestamp: now() } }
    }
    return {
      kind: 'Conflict',
      conflict: { localSnapshot: toSnapshot(localData), remoteSnapshot: toSnapshot(remoteData) },
    }
  }

  async resolveConflict(keepLocal: boolean): Promise<SyncResult> {
    if (keepLocal) {
      const localData = this.buildLocalSyncData()
      await this.cloudRepo.push(localData)
      return {
        pushed: localData.players.length + localData.gameTypes.length + localData.matches.length,
        pulled: 0,
        timestamp: now(),
      }
    }
    const remoteData = await this.cloudRepo.pull()
    this.writeRemoteToLocal(remoteData)
    return {
      pushed: 0,
      pulled: remoteData.players.length + remoteData.gameTypes.length + remoteData.matches.length,
      timestamp: now(),
    }
  }

  async login(): Promise<void> {
    await this.cloudRepo.login()
  }

  async logout(): Promise<void> {
    await this.cloudRepo.logout()
  }

  async status(): Promise<SyncStatus> {
    return this.cloudRepo.getStatus()
  }

  private buildLocalSyncData(): SyncData {
    return {
      players: this.playerRepo.getAll(true),
      gameTypes: this.gameTypeRepo.getAll(),
      matches: this.matchRepo.getAll(),
      lastModified: now(),
    }
  }

  private writeRemoteToLocal(data: SyncData): void {
    const localPlayers = this.playerRepo.getAll(true)
    const localGameTypes = this.gameTypeRepo.getAll(true)
    const localMatches = this.matchRepo.getAll()
    try {
      this.playerRepo.deleteAll()
      this.gameTypeRepo.deleteAll()
      this.matchRepo.deleteAll()
      this.playerRepo.saveAll(data.players)
      this.gameTypeRepo.saveAll(data.gameTypes)
      this.matchRepo.saveAll(data.matches)
    } catch (error) {
      this.playerRepo.deleteAll()
      this.gameTypeRepo.deleteAll()
      this.matchRepo.deleteAll()
      this.playerRepo.saveAll(localPlayers)
      this.gameTypeRepo.saveAll(localGameTypes)
      this.matchRepo.saveAll(localMatches)
      throw error
    }
  }
}
